FEC-10340 Multi-codec support for Audio/Video

GouravSna opened the pull request:
These APIs are part of `PlayerSettings`,

- `setPreferredVideoCodecSettings`: Set Preferred codec for video track. If the video has multiple codes (Ex: HEVC, AVC, AV1, VP9, VP8). 
`player.getSettings().setPreferredVideoCodecSettings(new VideoCodecSettings().setCodecPriorityList(Collections.singletonList(PKVideoCodec.HEVC)));
`

- `setPreferredAudioCodecSettings`: Set Preferred codec for audio track. If the video has multiple codes (Ex: AAC, AC3, E_AC3, OPUS).
`player.getSettings().setPreferredAudioCodecSettings(new AudioCodecSettings().setCodecPriorityList(Collections.singletonList(PKAudioCodec.AAC)));
`

	In both the above preferred codec Settings, you can pass your custom preferred list Singleton. There are other setters as well for Codec Settings,
	
	* `VideoCodecSettings` 
	
		`allowSoftwareDecoder` to allow the player to user software decoder if hardware decoder is not at all available, by default it is `false`.
	
		`allowMixedCodecAdaptiveness` if it is true then it will override your preferred codec list. It will add all the video codec in the tracks. If media has multiple codec available for video.
		
	* `AudioCodecSettings`
	
		`allowMixedCodecs` if it is true then it will override your preferred codec list. It will add all the audio codec in the tracks. If media has multiple codec available for audio.

GouravSna committed:
- Added support of HEVC, AVC with ABR

GouravSna committed:
- Code refactor

GouravSna committed:
- Check for codec is coming null
- Refactored the logic for ABR + Codec support

GouravSna committed:
- Checking the device support for HEVC

GouravSna committed:
- Added Javadoc

giladna committed:
use hashmap  for video tacks list  by codec

giladna committed:
fix build

giladna committed:
fix build - merge

giladna committed:
Merge branch 'dev' into FEM-2610_1

giladna committed:
no need to check twice
        if (uniqueIds.isEmpty()) {
            return;
        }

giladna committed:
no need in null check

giladna committed:
add isEmpty validation

GouravSna committed:
- Code refactor

giladna committed:
no need in exposing  public

giladna committed:
fix audio tracks AC3 EC3 and Adaptive

giladna committed:
add isDefaultStreamLang for filtering unneeded audio tracks

giladna committed:
update codec types

giladna committed:
upgrade  classpath 'com.novoda:bintray-release:0.9.1'

giladna committed:
Merge branch 'dev' into FEM-2610_1

# Conflicts:
#	playkit/src/main/java/com/kaltura/playkit/player/ExoPlayerWrapper.java
#	playkit/src/main/java/com/kaltura/playkit/player/TrackSelectionHelper.java

giladna committed:
add VideoCodecSettings
to configure Video Codec HEVC/AVC
and whether it exists on software decoder should it be used

player.getSettings().setPreferredVideoCodecSettings(new VideoCodecSettings().setSoftwareDecoderEnabled(true).setVideoCodec(PKVideoCodec.HEVC));

giladna committed:
Merge branch 'dev' into FEM-2610_1

# Conflicts:
#	playkit/src/main/java/com/kaltura/playkit/Player.java
#	playkit/src/main/java/com/kaltura/playkit/player/PlayerSettings.java

noamtamim committed:
Update ExoPlayerWrapper.java

noamtamim committed:
Update PlayerEngine.java

giladna committed:
add @NonNull

giladna committed:
support
    AV1,
    VP9,
    VP8,

giladna committed:
code review improovments

giladna committed:
clean imports

giladna committed:
add extra validation to play codes by priority

giladna committed:
using else if

giladna committed:
simplify code

giladna committed:
Merge branch 'dev' into FEM-2610_1

giladna committed:
roll back previous code change

giladna committed:
add comments

giladna committed:
Merge branch 'dev' into FEM-2610_1

# Conflicts:
#	playkit/src/main/java/com/kaltura/playkit/player/ExoPlayerWrapper.java
#	playkit/src/main/java/com/kaltura/playkit/player/TrackSelectionHelper.java

giladna committed:
Merge branch 'dev' into FEM-2610_1

# Conflicts:
#	playkit/src/main/java/com/kaltura/playkit/player/TrackSelectionHelper.java

accept mine

giladna committed:
update HEVC support implementation

giladna committed:
Merge branch 'dev' into FEM-2610_1

# Conflicts:
#	playkit/src/main/java/com/kaltura/playkit/Player.java
#	playkit/src/main/java/com/kaltura/playkit/player/AudioTrack.java
#	playkit/src/main/java/com/kaltura/playkit/player/ExoPlayerWrapper.java
#	playkit/src/main/java/com/kaltura/playkit/player/PlayerSettings.java
#	playkit/src/main/java/com/kaltura/playkit/player/TextTrack.java
#	playkit/src/main/java/com/kaltura/playkit/player/TrackSelectionHelper.java

giladna committed:
remove empty line

giladna committed:
remove empty line

giladna committed:
add missing NPE validation after code update

giladna committed:
Merge branch 'dev' into FEM-2610_1

giladna committed:
Merge branch 'dev' into FEM-2610_1

giladna committed:
Merge branch 'dev' into FEM-2610_1

# Conflicts:
#	playkit/src/main/java/com/kaltura/playkit/player/TrackSelectionHelper.java

giladna committed:
Merge branch 'dev' into FEM-2610_1

# Conflicts:
#	playkit/src/main/java/com/kaltura/playkit/player/ExoPlayerWrapper.java

giladna committed:
add  AudioCodecSettings for filtering by audio codec as well, so if manifest has 2 different codecs in 2 different groups  app can ask to have them both

giladna committed:
empty line

giladna committed:
Merge branch 'dev' into FEM-2610_1

* dev:
  - Fixed forceSinglePlayerEngine logic (#630)
  getProgramStartTime block execution incase of null player (#629)
  FEC-10107 Reject ABR if given values are not in the range coming in manifest (#628)
  FEC-10188 Add more info for UNEXPECTED and SOURCE_ERROR (#626)
  - Added AD_PROGRESS as class to get adPosition (#627)

giladna committed:
Merge branch 'dev' into FEM-2610_1

* dev:
  Compress device capabilities report before sending (#639)
  no need to clear the tracks in case of auto track is selected (#636)
  - Removed redundant file (#635)
  FEC-10202 File location refactor (#634)
  FEC-10202 Expose the IMA ErrorType ...  (#633)

giladna committed:
Merge branch 'dev' into FEM-2610_1

* dev:
  FEC-10315 - playback stuck if switch to invalid vtt track (404 response) and then seek (#643)
  making sure a window exists in index before getting it (#642)
  FEC-10294 - fix connection pool reuse issue on android 4.4 (#641)
  FEC-10212 ExoPlayer upgrade to 2.11.7 (#640)

GouravSna committed:
- null protections

GouravSna committed:
- Null Protection code refactor

GouravSna committed:
- Code refactor
- Added missing NPE with MissingResourceException

GouravSna committed:
- Removed space

noamtamim commented on playkit/src/main/java/com/kaltura/playkit/player/AudioCodecSettings.java:
```diff
@@ -0,0 +1,59 @@
+import java.util.List;
+
+public class AudioCodecSettings {
+    private List<PKAudioCodec> audioCodecPriorityList = new ArrayList<>();
+    private boolean allowAudioMixedMimeTypes = false;
```
Remove "audio" from field names
Change "MimeTypes" to "Codecs"

GouravSna replied:
Refactored the name

noamtamim commented on playkit/src/main/java/com/kaltura/playkit/player/VideoCodecSettings.java:
```diff
@@ -0,0 +1,72 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class VideoCodecSettings {
```
Remove "video" from all fields and functions

GouravSna replied:
done

noamtamim commented on playkit/src/main/java/com/kaltura/playkit/player/VideoCodecSettings.java:
```diff
@@ -0,0 +1,72 @@
+public class VideoCodecSettings {
+    private List<PKVideoCodec> videoCodecPriorityList = new ArrayList<>();
+    private boolean allowSoftwareDecoder = false;
+    private boolean allowVideoMixedMimeTypeAdaptiveness = false;
```
Rename "MimeType" to "Codec"

GouravSna replied:
Refactored the name

noamtamim commented on playkit/src/main/java/com/kaltura/playkit/player/TrackSelectionHelper.java:
```diff
@@ -254,10 +300,92 @@ private String getLanguageFromFormat(Format format) {
+            }
+        }
+
+        if (videoTracksCodecsMap.containsKey(PKVideoCodec.HEVC)) {
+            return  videoTracksCodecsMap.get(PKVideoCodec.HEVC);
+        } else if (videoTracksCodecsMap.containsKey(PKVideoCodec.AV1)) {
+            return  videoTracksCodecsMap.get(PKVideoCodec.AV1);
+        } else if (videoTracksCodecsMap.containsKey(PKVideoCodec.VP9)) {
+            return  videoTracksCodecsMap.get(PKVideoCodec.VP9);
+        } else if (videoTracksCodecsMap.containsKey(PKVideoCodec.VP8)) {
+            return  videoTracksCodecsMap.get(PKVideoCodec.VP8);
+        } else if (videoTracksCodecsMap.containsKey(PKVideoCodec.AVC)) {
+            return  videoTracksCodecsMap.get(PKVideoCodec.AVC);
+        } else {
+            return videoTracks;
+        }
```
For loop - create local array with all codecs by priority.

GouravSna replied:
Converted to forloop

noamtamim commented on playkit/src/main/java/com/kaltura/playkit/player/TrackSelectionHelper.java:
```diff
@@ -279,6 +407,21 @@ private String getLanguageFromFormat(Format format) {
+            return filteredAudioTracks;
+        }
+
+        if (audioTracksCodecsMap.containsKey(PKAudioCodec.E_AC3)) {
+            return  new ArrayList<>(audioTracksCodecsMap.get(PKAudioCodec.E_AC3));
+        } else if (audioTracksCodecsMap.containsKey(PKAudioCodec.AC3)) {
+            return  new ArrayList<>(audioTracksCodecsMap.get(PKAudioCodec.AC3));
+        } else if (audioTracksCodecsMap.containsKey(PKAudioCodec.OPUS)) {
+            return  new ArrayList<>(audioTracksCodecsMap.get(PKAudioCodec.OPUS));
+        } else if (audioTracksCodecsMap.containsKey(PKAudioCodec.AAC)) {
+            return  new ArrayList<>(audioTracksCodecsMap.get(PKAudioCodec.AAC));
+        }
```
For loop like in video.

GouravSna replied:
Converted to forloop

GouravSna committed:
- Removed non null annotation from Player Settings' interfaces
- Refactored the names of variables in Video/AudioCodecSettings
- Using for loop to get the values from map for filteration

GouravSna committed:
- Refactored the code.
- Removed nonnull from PlayerSettings
